File: Minimize the Heights II/sol.py
class Solution:
    def getMinDiff(self, arr, n, k):
        arr.sort()
        minimum = arr[0]
        maximum = arr[n - 1]
        diff = maximum - minimum

        for i in range(1, n):
            if arr[i] - k <= 0:
                continue
            
            new_min = min(arr[0] + k, arr[i] - k)
            new_max = max(arr[n - 1] - k, arr[i - 1] + k)
            diff = min(diff, new_max - new_min)

        return diff

File: Minimize the Heights II/sol.cpp
class Solution {
  public:
        int getMinDiff(int arr[], int n, int k) {
        sort(arr, arr+n);  
        int min2 = arr[0];
        int max2 = arr[n-1];
        int diff = max2 - min2;
        
        for(int i=1; i<n; i++){
            if(arr[i]-k <= 0){
                continue;
            }
            
            min2 = min(arr[0] + k, arr[i] - k);
            max2 = max(arr[n-1]-k, arr[i-1]+k);
            diff = min(diff, max2-min2);
        }
        
        return diff;
    }
    
};
